module1-introduction-to-backend/r3.4-api-best-practices/README.md

```diff
@@ -0,0 +1,232 @@
+
+The combination of the HTTP verb and resource noun convey the purpose of an API endpoint. API endpoints like GET `/fetchArticles` or POST `/update-articles` use redundant verbs like "fetch" and "update" which is not required.
+
+In Express, we can write these endpoints as:
+```
+const express = require('express');
+const bodyParser = require('body-parser');
+
+const app = express();
+
+app.use(bodyParser.json());
+
+app.get('/articles', (req, res) => {
+  const articles = [];
+  // code to retrieve an article...
+  res.json(articles);
+});
+
+app.post('/articles', (req, res) => {
+  // code to add a new article...
+  res.json(req.body);
+});
+
+app.put('/articles/:id', (req, res) => {
+  const { id } = req.params;
+  // code to update an article...
+  res.json(req.body);
+});
+
+app.delete('/articles/:id', (req, res) => {
+  const { id } = req.params;
+  // code to delete an article...
+  res.json({ deleted: id });
+});
+
+app.listen(3000, () => console.log('server started'));
+```
+
+3. **Use logical nesting on endpoints**
+When designing endpoints, it makes sense to group those that contain associated information. That is, if one object can contain another object, you should design the endpoint to reflect that. This is good practice regardless of whether your data is structured like this in your database. In fact, it may be advisable to avoid mirroring your database structure in your endpoints to avoid giving attackers unnecessary information.
+
+For example, if we want an endpoint to get the comments for a news article, we should append the `/comments` path to the end of the `/articles` path.
+
+In Express, we can write this as:
+```
+const express = require('express');
+const bodyParser = require('body-parser');
+
+const app = express();
+
+app.use(bodyParser.json());
+
+app.get('/articles/:articleId/comments', (req, res) => {
+  const { articleId } = req.params;
+  const comments = [];
+  // code to get comments by articleId
+  res.json(comments);
+});
+
+
+app.listen(3000, () => console.log('server started'));
+```
+
+4. **Handle complexity elegantly**
+The data you’re trying to expose can be characterized by a lot of properties which could be beneficial for the end consumer working with your API. These properties describe the base resource and isolate specific assets of information that can be manipulated with the appropriate method. An API should strive towards completion, and provide all the required information, data and resources to help developers integrate with them in a seamless manner.
+
+However, instead of defining more resources and endpoints to cover dynamic use cases and relationships, you can sweep properties and limit responses behind the ‘?’ in a query parameter, or isolate specific component of the data the client is working with using a path parameter.
```

Comment:
perhaps emphasize there is no right or wrong answer here but you should consider what would be cleanest in terms of the user perspective, and sometimes either option could be ok. or in general, i imagine a student asking "when do i add a query parameter vs using a route"?